src/game.cpp

```diff
@@ -10242,245 +10242,226 @@ void game::grab()
+        } else if( u.pour_into( *cont, liquid ) ) {
+            u.mod_moves( -100 );
         }
         return true;
+    }
 
+    uimenu menu;
+    menu.return_invalid = true;
+    if( source_pos != nullptr ) {
+        menu.text = string_format( _( "What to do with the %s from %s?" ), liquid_name.c_str(), m.name( *source_pos ).c_str() );
+    } else if( source_veh != nullptr ) {
+        menu.text = string_format( _( "What to do with the %s from the %s?" ), liquid_name.c_str(), source_veh->name.c_str() );
     } else {
-        // Filling up normal containers
-        bool allow_bucket = cont == &u.weapon || !u.has_item( *cont );
-        std::string err;
-        if( !cont->fill_with( liquid, err, allow_bucket ) ) {
-            add_msg( m_info, err.c_str() );
-            return false;
-        }
-
-        u.inv.unsort();
-        add_msg( _( "You pour %1$s into the %2$s." ), liquid.tname().c_str(), cont->tname().c_str() );
-        if( !infinite && liquid.charges > 0 ) {
-            add_msg( _( "There's some left over!" ) );
-        }
-        return infinite || liquid.charges <= 0;
+        menu.text = string_format( _( "What to do with the %s?" ), liquid_name.c_str() );
     }
+    std::vector<std::function<void()>> actions;
 
-    return false;
-}
-
-//Move_liquid returns the amount of liquid left if we didn't move all the liquid, otherwise returns sentinel -1, signifies transaction fail.
-//One-use, strictly for liquid transactions. Not intended for use with while loops.
-int game::move_liquid(item &liquid)
-{
-    if (!liquid.made_of(LIQUID)) {
-        dbg(D_ERROR) << "game:move_liquid: Tried to move_liquid a non-liquid!";
-        debugmsg("Tried to move_liquid a non-liquid!");
-        return -1;
+    if( liquid.is_food( &u ) ) {
+        menu.addentry( -1, true, 'e', _( "Consume it" ) );
+        actions.emplace_back( [&]() {
+            // consume_item already consumes moves.
+            u.consume_item( liquid );
+        } );
     }
 
-    //liquid is in fact a liquid.
-    const std::string text = string_format(_("Container for %s"), liquid.tname().c_str());
-    int pos = inv_for_liquid(liquid, text, false);
+    for( auto &veh : nearby_vehicles_for( liquid.typeId() ) ) {
+        if( veh == source_veh ) {
+            continue;
+        }
+        menu.addentry( -1, true, MENU_AUTOASSIGN, _( "Fill nearby vehicle %s" ), veh->name.c_str() );
+        actions.emplace_back( [&, veh]() {
+            if( create_activity() ) {
+                serialize_liquid_target( u.activity, *veh );
+                return;
+            }
+            if( u.pour_into( *veh, liquid ) ) {
+                u.mod_moves( -100 );
+            }
+        } );
+    }
 
-    //is container selected?
-    item *cont = &( u.i_at( pos ) );
-    if( u.has_item( *cont ) ) {
-        if (cont == NULL || cont->is_null()) {
-            return -1;
+    menu.addentry( -1, true, 'g', _( "Pour on the ground / into an adjacent keg" ) );
```

Comment:
So you propose another menu entry "Pour into adjacent keg"? Or a separate entry for *each* adjacent keg ("Pour into keg (Northwest)", "Pour into keg (North)" - like it's done for the vehicles?